File: app/src/main/java/com/ilyap/yuta/ui/dialogs/team/EditTeamDialog.java
package com.ilyap.yuta.ui.dialogs.team;

import android.content.Context;
import android.widget.TextView;
import androidx.fragment.app.Fragment;
import com.ilyap.yuta.R;
import com.ilyap.yutarefactor.domain.entity.Team;
import com.ilyap.yutarefactor.domain.response.TeamResponse;
import com.ilyap.yutarefactor.domain.response.UpdateResponse;
import com.ilyap.yuta.ui.fragments.TeamsFragment;
import com.ilyap.yuta.utils.RequestViewModel;

@SuppressWarnings("ConstantConditions")
public class EditTeamDialog extends CreateTeamDialog {
    private final int teamId;
    private Team team;

    public EditTeamDialog(Context context, Fragment fragment, int teamId) {
        super(context, fragment);
        this.teamId = teamId;
    }

    @Override
    public void start() {
        super.start();
        setupViews();
        getTeam(teamId);

        submitButton.setEnabled(false);
        submitButton.setOnClickListener(v -> {
            hideKeyboard(submitButton);
            editTeam();
        });
    }

    private void setupViews() {
        submitButton.setText(R.string.save_button);
        ((TextView) dialog.findViewById(R.id.create_text)).setText(getContext().getString(R.string.edit_team_text));
    }

    private void getTeam(int teamId) {
        viewModel.getResultLiveData().removeObservers(fragment);
        viewModel.getTeam(teamId);
        viewModel.getResultLiveData().observe(fragment, result -> {
            if (!(result instanceof TeamResponse)) return;
            team = ((TeamResponse) result).getTeam();
            setupTeam();
        });
    }

    private void setupTeam() {
        teamName.setText(team.getName());
        membersAdapter.updateList(team.getMembers());
        updateAddedTextVisibility();
    }

    private void editTeam() {
        viewModel.getResultLiveData().removeObservers(fragment);
        viewModel.editTeam(teamId, getData(teamName), addedMembers);
        viewModel.getResultLiveData().observe(fragment, result -> {
            if (!(result instanceof UpdateResponse)) return;
            ((TeamsFragment) fragment).updateLists();
            dismiss();
        });
    }

    @Override
    protected RequestViewModel isNameUnique(String name) {
        viewModel.getResultLiveData().removeObservers(fragment);
        viewModel.checkUniqueTeamName(name, teamId);
        return viewModel;
    }
}

File: app/src/main/java/com/yuta/teams/ui/dialog/EditTeamDialog.kt
package com.yuta.teams.ui.dialog

import android.widget.TextView
import androidx.fragment.app.Fragment
import androidx.fragment.app.viewModels
import androidx.lifecycle.viewModelScope
import com.yuta.app.R
import com.yuta.common.util.FieldUtils.trimmedText
import com.yuta.common.util.KeyboardUtils
import com.yuta.domain.model.Team
import com.yuta.domain.model.UserDto
import com.yuta.teams.viewmodel.TeamDialogsViewModel
import kotlinx.coroutines.flow.first
import kotlinx.coroutines.launch

class EditTeamDialog(
    private val fragment: Fragment,
    private val team: Team,
    private val onEditSuccessCallback: () -> Unit
) : CreateTeamDialog(fragment) {

    private val title: TextView by lazy { dialog.findViewById(R.id.title) }

    private val teamViewModel: TeamDialogsViewModel by fragment.viewModels()

    override fun start() {
        super.start()
        setupViews()
        populateTeamDetails(team)

        submitButton.setOnClickListener {
            KeyboardUtils.hideKeyboard(fragment.requireActivity(), submitButton)
            editTeam(team.id, teamName.trimmedText(), teamViewModel.addedMembers)
        }
    }

    private fun setupViews() {
        submitButton.apply {
            setText(R.string.save_button)
            isEnabled = false
        }
        title.setText(R.string.edit_team_text)
    }

    private fun populateTeamDetails(team: Team) {
        teamName.setText(team.name)
        membersAdapter.refillList(team.members)
        updateAddedTextVisibility()
    }

    private fun editTeam(teamId: Int, name: String, members: List<UserDto>) {
        teamViewModel.viewModelScope.launch {
            val isEdited = teamViewModel.edit(teamId, name, members).first()
            handleEditResult(isEdited)
        }
    }

    private fun handleEditResult(isEdited: Boolean) {
        if (isEdited) {
            onEditSuccessCallback()
            dismiss()
        }
    }

    override fun checkNameUnique(name: String, onUniqueCallback: (Boolean) -> Unit) {
        teamViewModel.viewModelScope.launch {
            teamViewModel.isUniqueName(name, team.id).collect { onUniqueCallback(it) }
        }
    }
}
